### PR TITLE
Update cosign verify signed Kubernetes artifacts for 2.0

### LINKED ISSUE #39775: Update page Verify Signed Kubernetes Artifacts for cosign 2.0
**This is a Feature Request**
After [cosign 2.0](https://blog.sigstore.dev/cosign-2-0-released/) release, we don't need `COSIGN_EXPERIMENTAL=1`. We should update the instructions on the task [page](https://kubernetes.io/docs/tasks/administer-cluster/verify-signed-artifacts/) to include verification using cosign 2.0

**What would you like to be added**
<details>
<summary> verification using cosign 2.0</summary>

```bash
~ $ cosign verify registry.k8s.io/kube-apiserver:v1.26.0 --certificate-identity-regexp krel-trust@k8s-releng-prod.iam.gserviceaccount.com --certificate-oidc-issuer https://accounts.google.com | jq .

Verification for registry.k8s.io/kube-apiserver:v1.26.0 --
The following checks were performed on each of these signatures:
  - The cosign claims were validated
  - Existence of the claims in the transparency log was verified offline
  - The code-signing certificate was verified using trusted certificate authority certificates
[
  {
    "critical": {
      "identity": {
        "docker-reference": "europe-west1-docker.pkg.dev/k8s-artifacts-prod/images/kubernetes/kube-apiserver"
      },
      "image": {
        "docker-manifest-digest": "sha256:d230a0b88a3daf14e4cce03b906b992c8153f37da878677f434b1af8c4e8cc75"
      },
      "type": "cosign container image signature"
    },
    "optional": {
      "1.3.6.1.4.1.57264.1.1": "https://accounts.google.com",
      "Bundle": {
        "SignedEntryTimestamp": "MEYCIQDnvT+cu60FogfZIdroUCP3Mrckp9aILhcvFik9k+NoLwIhAMoZ+shE4tj4SNbN+a29YfAmloCm3S8tZ3jCDbC4o+/B",
        "Payload": {
          "body": "eyJhcGlWZXJzaW9uIjoiMC4wLjEiLCJraW5kIjoiaGFzaGVkcmVrb3JkIiwic3BlYyI6eyJkYXRhIjp7Imhhc2giOnsiYWxnb3JpdGhtIjoic2hhMjU2IiwidmFsdWUiOiI5NGYyZTZlY2FiNTY3YzcwMjQxYjNhN2M5ZTNiMzZlYjg1ZGFkOTEzZDY5M2M5MTgzNmY3MGJkNTNhZjljODI5In19LCJzaWduYXR1cmUiOnsiY29udGVudCI6Ik1FVUNJUUM2b1BTbSs1cldHY2ptTnR0eTR6dHVUeVBMRnM5dlk5eFZVUmswaWZjWENRSWdRdURQOGlxbzB2QysweGxHTm11QU5VUGx1QVoxZEpSd0FGRzlxdXY2RlY4PSIsInB1YmxpY0tleSI6eyJjb250ZW50IjoiTFMwdExTMUNSVWRKVGlCRFJWSlVTVVpKUTBGVVJTMHRMUzB0Q2sxSlNVTjJla05EUVd0VFowRjNTVUpCWjBsVlIxSkZhVXhqVDJjdlYySlFNekkwUlhodFRsTk9Xamt6TlhaemQwTm5XVWxMYjFwSmVtb3dSVUYzVFhjS1RucEZWazFDVFVkQk1WVkZRMmhOVFdNeWJHNWpNMUoyWTIxVmRWcEhWakpOVWpSM1NFRlpSRlpSVVVSRmVGWjZZVmRrZW1SSE9YbGFVekZ3WW01U2JBcGpiVEZzV2tkc2FHUkhWWGRJYUdOT1RXcEplRTFxUVRSTmFrMTVUMFJSZVZkb1kwNU5ha2w0VFdwQk5FMXFUWHBQUkZGNVYycEJRVTFHYTNkRmQxbElDa3R2V2tsNmFqQkRRVkZaU1V0dldrbDZhakJFUVZGalJGRm5RVVV6VEZGT1VsSkdXREJPY1UwMFUyRlBlVWhuWWxCNmVrWXdTVEF5VkRWUldrTkpNMEVLZVdkaGNIaFRlVU5JYUVWa1F5OVljV3BxZEU5MlZXdFFRamRYUjNwYVF6QmpaMUZOZFhaM1pIWnZURFpXTVVoNFNYRlBRMEZYVFhkblowWm1UVUUwUndwQk1WVmtSSGRGUWk5M1VVVkJkMGxJWjBSQlZFSm5UbFpJVTFWRlJFUkJTMEpuWjNKQ1owVkdRbEZqUkVGNlFXUkNaMDVXU0ZFMFJVWm5VVlZpYUU5NENrMW1WRlJxZEVwMk4wdHNaRmRUUkVoNFJYaG1TVUZKZDBoM1dVUldVakJxUWtKbmQwWnZRVlV6T1ZCd2VqRlphMFZhWWpWeFRtcHdTMFpYYVhocE5Ga0tXa1E0ZDFGQldVUldVakJTUVZGSUwwSkVXWGRPU1VWNVlUTktiR0pETVRCamJsWjZaRVZDY2s5SVRYUmpiVlp6V2xjMWJreFlRbmxpTWxGMVlWZEdkQXBNYldSNldsaEtNbUZYVG14WlYwNXFZak5XZFdSRE5XcGlNakIzUzFGWlMwdDNXVUpDUVVkRWRucEJRa0ZSVVdKaFNGSXdZMGhOTmt4NU9XaFpNazUyQ21SWE5UQmplVFZ1WWpJNWJtSkhWWFZaTWpsMFRVbEhTMEpuYjNKQ1owVkZRV1JhTlVGblVVTkNTSGRGWldkQ05FRklXVUV6VkRCM1lYTmlTRVZVU21vS1IxSTBZMjFYWXpOQmNVcExXSEpxWlZCTE15OW9OSEI1WjBNNGNEZHZORUZCUVVkRk9VSklabFJuUVVGQ1FVMUJVbnBDUmtGcFJVRjVhRlZKUzNkUVJRcHVaMEZTYlZGeU5tMTZlazlpZERkalpESXhhMjV2TjFSaGEycDVVMUlyYXpGRk1FTkpRMlp0TDJ4elQwaDNjM05sVWpod1oyUm9VRWRHZW1GaVZVTjZDakpNUXpBNVRuUmpVM1JSYlRKWWFXVk5RVzlIUTBOeFIxTk5ORGxDUVUxRVFUSnJRVTFIV1VOTlVVTlJkRWRQZWs1YUswWldkU3RpUXk4d1FXVm9SSFlLUmtaclMya3liMkZNVldkMWVHeDVaMHN2YmtKUVZtcE9aakkxZVVoWlFXazNXVTF2Ym01UFQzRXdORU5OVVVRck5XVnVibEJpVUZWNGVrTkVhMGRqZEFwT1YyNTFUVEpSZVZVclVXUmhSRkJEWW1oYWFHTkxUelZRWjNSRVJUaDRWRkZ5TTBRNVprVTFPVU4yWmpFNWJ6MEtMUzB0TFMxRlRrUWdRMFZTVkVsR1NVTkJWRVV0TFMwdExRbz0ifX19fQ==",
          "integratedTime": 1670542146,
          "logIndex": 8691003,
          "logID": "c0d23d6ad406973f9559f3ba2d1ca01f84147d8ffc5b8445c224f98b9591801d"
        }
      },
      "Issuer": "https://accounts.google.com",
      "Subject": "krel-trust@k8s-releng-prod.iam.gserviceaccount.com",
      "org.kubernetes.kpromo.mirrors": "asia-northeast2-docker.pkg.dev/k8s-artifacts-prod/images/kubernetes/kube-controller-manager-amd64,us-west1-docker.pkg.dev/k8s-artifacts-prod/images/kubernetes/kube-controller-manager-amd64,us.gcr.io/k8s-artifacts-prod/kubernetes/kube-controller-manager-amd64,us-south1-docker.pkg.dev/k8s-artifacts-prod/images/kube-controller-manager-amd64,us-west2-docker.pkg.dev/k8s-artifacts-prod/images/kubernetes/kube-controller-manager-amd64,europe-west9-docker.pkg.dev/k8s-artifacts-prod/images/kubernetes/kube-controller-manager-amd64,us-west1-docker.pkg.dev/k8s-artifacts-prod/images/kube-controller-manager-amd64,asia-northeast1-docker.pkg.dev/k8s-artifacts-prod/images/kube-controller-manager-amd64,europe-north1-docker.pkg.dev/k8s-artifacts-prod/images/kube-controller-manager-amd64,europe-southwest1-docker.pkg.dev/k8s-artifacts-prod/images/kubernetes/kube-controller-manager-amd64,asia-east1-docker.pkg.dev/k8s-artifacts-prod/images/kubernetes/kube-controller-manager-amd64,australia-southeast1-docker.pkg.dev/k8s-artifacts-prod/images/kubernetes/kube-controller-manager-amd64,us-east4-docker.pkg.dev/k8s-artifacts-prod/images/kubernetes/kube-controller-manager-amd64,asia-northeast2-docker.pkg.dev/k8s-artifacts-prod/images/kube-controller-manager-amd64,us-south1-docker.pkg.dev/k8s-artifacts-prod/images/kubernetes/kube-controller-manager-amd64,europe-west1-docker.pkg.dev/k8s-artifacts-prod/images/kubernetes/kube-controller-manager-amd64,asia.gcr.io/k8s-artifacts-prod/kubernetes/kube-controller-manager-amd64,us-central1-docker.pkg.dev/k8s-artifacts-prod/images/kube-controller-manager-amd64,asia-northeast1-docker.pkg.dev/k8s-artifacts-prod/images/kubernetes/kube-controller-manager-amd64,us-east1-docker.pkg.dev/k8s-artifacts-prod/images/kubernetes/kube-controller-manager-amd64,asia.gcr.io/k8s-artifacts-prod/kube-controller-manager-amd64,europe-west2-docker.pkg.dev/k8s-artifacts-prod/images/kube-controller-manager-amd64,southamerica-west1-docker.pkg.dev/k8s-artifacts-prod/images/kube-controller-manager-amd64,europe-west8-docker.pkg.dev/k8s-artifacts-prod/images/kubernetes/kube-controller-manager-amd64,us-east5-docker.pkg.dev/k8s-artifacts-prod/images/kube-controller-manager-amd64,asia-south1-docker.pkg.dev/k8s-artifacts-prod/images/kube-controller-manager-amd64,us-east4-docker.pkg.dev/k8s-artifacts-prod/images/kube-controller-manager-amd64,us-west2-docker.pkg.dev/k8s-artifacts-prod/images/kube-controller-manager-amd64,eu.gcr.io/k8s-artifacts-prod/kube-controller-manager-amd64,europe-north1-docker.pkg.dev/k8s-artifacts-prod/images/kubernetes/kube-controller-manager-amd64,southamerica-west1-docker.pkg.dev/k8s-artifacts-prod/images/kubernetes/kube-controller-manager-amd64,europe-west4-docker.pkg.dev/k8s-artifacts-prod/images/kubernetes/kube-controller-manager-amd64,eu.gcr.io/k8s-artifacts-prod/kubernetes/kube-controller-manager-amd64,us.gcr.io/k8s-artifacts-prod/kube-controller-manager-amd64,europe-west8-docker.pkg.dev/k8s-artifacts-prod/images/kube-controller-manager-amd64,asia-south1-docker.pkg.dev/k8s-artifacts-prod/images/kubernetes/kube-controller-manager-amd64,europe-west9-docker.pkg.dev/k8s-artifacts-prod/images/kube-controller-manager-amd64,australia-southeast1-docker.pkg.dev/k8s-artifacts-prod/images/kube-controller-manager-amd64,us-central1-docker.pkg.dev/k8s-artifacts-prod/images/kubernetes/kube-controller-manager-amd64,asia-east1-docker.pkg.dev/k8s-artifacts-prod/images/kube-controller-manager-amd64,us-east5-docker.pkg.dev/k8s-artifacts-prod/images/kubernetes/kube-controller-manager-amd64,europe-west2-docker.pkg.dev/k8s-artifacts-prod/images/kubernetes/kube-controller-manager-amd64,europe-southwest1-docker.pkg.dev/k8s-artifacts-prod/images/kube-controller-manager-amd64,europe-west4-docker.pkg.dev/k8s-artifacts-prod/images/kube-controller-manager-amd64,us-east1-docker.pkg.dev/k8s-artifacts-prod/images/kube-controller-manager-amd64,europe-west1-docker.pkg.dev/k8s-artifacts-prod/images/kube-controller-manager-amd64"
    }
  }
]
```
Add a note if cosign version < 2.0
</details>

**Why is this needed**
There are some breaking changes in cosign 2.0 and we should update the instruction for the verification of images. 
**Comments**

### PR DESCRIPTION
**Pull request by mrgiles**
Update **Verify Signed Kubernetes Artifacts** instructions to work with Cosign 2.0. Fixes #39775

### CODE REVIEW

**Comment by mrgiles** [comment]
**Note:** I had to remove the **Verifying images for all control plane components** entry because the commands that are shown in it (`curl -Ls https://sbom.k8s.io/$(curl -Ls https://dl.k8s.io/release/latest.txt)/release | grep 'PackageName: registry.k8s.io/' | awk '{print $2}'`) don't seem to work at all, at least in my tests. Perhaps this could be due to changes in the contents of both `sbom.k8s.io` and `registry.k8s.io`?

If you think this entry is important, I welcome suggestions on how to fix those commands to make them produce the desired output, a complete list of all the control plane images stored on `registry.k8s.io`. I tried several iterations of those commands but none worked for me.

**Comment by mrgiles** [comment]
FYI @kranurag7

File: content/en/docs/tasks/administer-cluster/verify-signed-artifacts.md [hunk 1]
```diff
@@ -20,15 +20,21 @@ You will need to have the following tools installed:
 - `cosign` ([install guide](https://docs.sigstore.dev/cosign/installation/))
 - `curl` (often provided by your operating system)
 
+{{< note >}}
+`COSIGN_EXPERIMENTAL=1` is no longer required to have identity-based ("keyless") signing and transparency.
+
+For additional information please read the Cosign 2.0 [blog post](https://blog.sigstore.dev/cosign-2-0-released/).
+{{< /note >}}
```
Comment (sftim) [inline content/en/docs/tasks/administer-cluster/verify-signed-artifacts.md:27]: Can you rephrase this to be [timeless](https://developers.google.com/style/timeless-documentation)?

For example, start with: “Previous versions of Cosign required that you enabled the `COSIGN_EXPERIMENTAL`&hellip;”

File: content/en/docs/tasks/administer-cluster/verify-signed-artifacts.md [hunk 2]
```diff
@@ -20,15 +20,21 @@ You will need to have the following tools installed:
...
 The Kubernetes release process signs all binary artifacts (tarballs, SPDX files,
-standalone binaries) by using cosign's keyless signing. To verify a particular
-binary, retrieve it together with its signature and certificate:
+standalone binaries) by using cosign's [Keyless signing](https://docs.sigstore.dev/cosign/keyless/). To verify a particular
+binary, retrieve it together with its signature and certificate files:
 
 ```bash
 URL=https://dl.k8s.io/release/v{{< skew currentVersion >}}.0/bin/linux/amd64
-BINARY=kubectl
+BINARY=kube-apiserver
```
Comment (sftim) [inline content/en/docs/tasks/administer-cluster/verify-signed-artifacts.md:37]: Why change the binary? Not many people download the API server directly.

File: content/en/docs/tasks/administer-cluster/verify-signed-artifacts.md [hunk 3]
```diff
@@ -41,58 +47,37 @@ for FILE in "${FILES[@]}"; do
 done
 ```
 
-Then verify the blob by using `cosign`:
+Then verify the blob using `cosign verify-blob`:
 
 ```shell
-cosign verify-blob "$BINARY" --signature "$BINARY".sig --certificate "$BINARY".cert
+cosign verify-blob "$BINARY" --signature "$BINARY".sig --certificate "$BINARY".cert --certificate-identity krel-staging@k8s-releng-prod.iam.gserviceaccount.com --certificate-oidc-issuer https://accounts.google.com
```
Comment (sftim) [inline content/en/docs/tasks/administer-cluster/verify-signed-artifacts.md:53]: > `krel-staging@k8s-releng-prod.iam.gserviceaccount.com`

This will need tech review to confirm that we want to codify that identity.

File: content/en/docs/tasks/administer-cluster/verify-signed-artifacts.md [hunk 4]
```diff
@@ -41,58 +47,37 @@ for FILE in "${FILES[@]}"; do
...
-`--cert` for older versions of cosign.
+A succesfull verification will output `Verified OK`.
 
 {{< note >}}
-To learn more about keyless signing, please refer to [Keyless
-Signatures](https://github.com/sigstore/cosign/blob/main/KEYLESS.md#keyless-signatures).
+Verification now requires identity flags, `--certificate-identity` and `--certificate-oidc-issuer`.
+
+For additional information please read the Cosign 2.0 [blog post](https://blog.sigstore.dev/cosign-2-0-released/).
 {{< /note >}}
```
Comment (sftim) [inline content/en/docs/tasks/administer-cluster/verify-signed-artifacts.md:62]:
```suggestion
```

Omit all this.

File: content/en/docs/tasks/administer-cluster/verify-signed-artifacts.md [hunk 5]
```diff
@@ -41,58 +47,37 @@ for FILE in "${FILES[@]}"; do
...
+Then verify the blob using `cosign verify-blob`:
 
 ```shell
-cosign verify-blob "$BINARY" --signature "$BINARY".sig --certificate "$BINARY".cert
+cosign verify-blob "$BINARY" --signature "$BINARY".sig --certificate "$BINARY".cert --certificate-identity krel-staging@k8s-releng-prod.iam.gserviceaccount.com --certificate-oidc-issuer https://accounts.google.com
 ```
 
-cosign v1.9.0 is required to be able to use the `--certificate` flag. Please use
-`--cert` for older versions of cosign.
+A succesfull verification will output `Verified OK`.
```
Comment (sftim) [inline content/en/docs/tasks/administer-cluster/verify-signed-artifacts.md:56]:
```suggestion
A succesful verification will output `Verified OK`.
```
We prefer to put outputs into multiline code blocks, though.

File: content/en/docs/tasks/administer-cluster/verify-signed-artifacts.md [hunk 6]
```diff
@@ -41,58 +47,37 @@ for FILE in "${FILES[@]}"; do
...
+cosign verify registry.k8s.io/kube-apiserver-amd64:v{{< skew currentVersion >}}.0 --certificate-identity krel-trust@k8s-releng-prod.iam.gserviceaccount.com --certificate-oidc-issuer https://accounts.google.com
 ```
 
-{{< note >}}
-`COSIGN_EXPERIMENTAL=1` is used to allow verification of images signed
-in `KEYLESS` mode. To learn more about keyless signing, please refer to
-[Keyless Signatures](https://github.com/sigstore/cosign/blob/main/KEYLESS.md#keyless-signatures)
-. {{< /note >}}
-
-### Verifying images for all control plane components
```
Comment (sftim) [inline content/en/docs/tasks/administer-cluster/verify-signed-artifacts.md:76]: Why remove this?

**Review by sftim — COMMENTED** [review]
Thanks

/sig release
/sig security

This will need tech review.

File: content/en/docs/tasks/administer-cluster/verify-signed-artifacts.md (hunk 6 above)
Comment (mrgiles) [reply content/en/docs/tasks/administer-cluster/verify-signed-artifacts.md:76]: I've explained why I removed the entry in a [previous comment](https://github.com/kubernetes/website/pull/40910#issuecomment-1529142829). Those commands don't work as written.I tried different alternatives but none of them produced useful output

Comment (sftim) [reply content/en/docs/tasks/administer-cluster/verify-signed-artifacts.md:76]: /hold

We should find another fix. Ideally, we find a way to retain that heading and provide correct (or at least correct-for-the-alpha) advice.

It's OK to unhold this if SIG Release confirm that it's not possible for an end user to verify container image artifacts are genuine.

Comment (sftim) [reply content/en/docs/tasks/administer-cluster/verify-signed-artifacts.md:76]: There's a related concern: the PR short heading doesn't mention or hint at this removal.

Comment (mrgiles) [reply content/en/docs/tasks/administer-cluster/verify-signed-artifacts.md:76]: An end user CAN verify if control plane container images are genuine using `cosign`, but providing a list of image names in a prepopulated file. What I wasn't able to do is to automatically produce a list of all the control plane images using the command provided in the document:

```
curl -Ls https://sbom.k8s.io/$(curl -Ls https://dl.k8s.io/release/latest.txt)/release | grep 'PackageName: registry.k8s.io/' | awk '{print $2}' > images.txt
input=images.txt
while IFS= read -r image
do
  COSIGN_EXPERIMENTAL=1 cosign verify "$image"
done < "$input"
```

I think that the content of the registries has changed, probably this script was written when the images were stored on `k8s.gcr.io` but now on `k8s.gcr.io`, I think that images are also stored on AWS S3 buckets as well and they are probably using more complex redirects, so my guess is that those changes are breaking the functionality of this script.

As you said, getting input from the teams directly involved in these areas would be good.

ACK, if content is removed, it should be clearly stated in the PR comment.

Thanks!

Comment (mrgiles) [reply content/en/docs/tasks/administer-cluster/verify-signed-artifacts.md:76]: I have a working prototype shell script, it's a lot longer than the original but it works. I'm sure it can be improved but this is what I have for now:

```shell
#!/bin/bash

# Control plane image names
images=(kube-apiserver kube-controller-manager kube-scheduler kube-proxy conformance)

# Get the latest stable image version from dl.k8s.io/release/stable.txt
version=$(curl -Ls https://dl.k8s.io/release/stable.txt)

#  Image architecture name
arch=$1

# Require the image architecture name
[ $# -eq 0 ] && { echo "missing arch name [amd64|arm64|ppc64le|s390x]"; exit 1; }

# Loop through the images and verify them using cosign
for image in "${images[@]}"; do
  # Get the image names from registry.k8s.io
  name=$(curl -s https://registry.k8s.io/v2/${image}/tags/list | grep -o "\"${arch}-v[0-9.]*\"" | awk -F\" '{print $2}')
  
  # Verify each image with cosign using registry/name-arch:version --certificate-identity --certificate-oidc-issuer
  echo "===> Verifiying image: registry.k8s.io/${image}${name}-${arch}:${version} <==="
  cosign verify "registry.k8s.io/${image}${name}-${arch}:${version}" --certificate-identity krel-trust@k8s-releng-prod.iam.gserviceaccount.com --certificate-oidc-issuer https://accounts.google.com | jq -r .
done
```

This another option, which is the closest I could get to the original script, but it produces a list of images with lots of errors at the time of verifying them, like they either don't exist or the name format is incorrect, not sure:

```shell
arch="amd64"
version=$(curl -Ls https://dl.k8s.io/release/stable.txt)
curl -Ls "https://sbom.k8s.io/${version}/source" | grep 'PackageName: k8s.io/' | awk '{print $2}' > images.txt
input=images.txt
while IFS= read -r image
do
  cosign verify "registry.${image}-${arch}:${version}" --certificate-identity krel-trust@k8s-releng-prod.iam.gserviceaccount.com --certificate-oidc-issuer https://accounts.google.com | jq -r .
done < "$input"
```

Suggestions welcomed. Thanks!

FYI: @kranurag7 @sftim

Comment (mrgiles) [reply content/en/docs/tasks/administer-cluster/verify-signed-artifacts.md:76]: Update: I think that Rijul Gulati in PR #40982 found the right combination of `curl+grep+cut+sed` to retrieve the names of the control plane images that need to be verified with cosign. So, IMO, the winning command would be:

```
curl -Ls "https://sbom.k8s.io/$(curl -Ls https://dl.k8s.io/release/stable.txt)/release" | grep "SPDXID: SPDXRef-Package-registry.k8s.io" |  grep -v sha256 | cut -d- -f3- | sed 's/-/\//' | sed 's/-v1/:v1/' > images.txt
input=images.txt
while IFS= read -r image
do
  cosign verify "$image" --certificate-identity krel-trust@k8s-releng-prod.iam.gserviceaccount.com --certificate-oidc-issuer https://accounts.google.com
done < "$input"
```

What is missing here are the values for `--certificate-identity` and `--certificate-oidc-issuer` for image and for blob. What should we use for these.

For the verify-blob command, the right arguments are:

`cosign verify-blob "$BINARY" --signature "$BINARY".sig --certificate "$BINARY".cert --certificate-identity krel-staging@k8s-releng-prod.iam.gserviceaccount.com --certificate-oidc-issuer https://accounts.google.com`

Result is:
`Verified OK`


The variations are `--certificate-identity krel-trust@k8s-releng-prod.iam.gserviceaccount.com` for `verify` and `--certificate-identity krel-staging@k8s-releng-prod.iam.gserviceaccount.com` for `verify-blob` respectively.

If all this can be settled, I can close this PR and have Rijul proceed with his (much cleaner) solution but adding these new required options for Cosign 2.0. But I'll leave that to the experts, these are just my 0.2 cents... 

FYI @sftim @PushkarJ @dipesh-rawat

Thanks!
